File: c++/Greedy.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <cstdlib>
#include <ctime>
#include <limits>

using namespace std;

// Función de distancia
double distancia(pair<int, int> p1, pair<int, int> p2) {
    return sqrt(pow(p1.first - p2.first, 2) + pow(p1.second - p2.second, 2));
}

// Función para calcular el largo de un recorrido
double largo_recorrido(vector<int>& ruta, vector<pair<int, int>>& puntos) {
    double total = 0;
    for (int i = 0; i < ruta.size() - 1; ++i) {
        total += distancia(puntos[ruta[i]], puntos[ruta[i + 1]]);
    }
    return total;
}

// Algoritmo Greedy
pair<vector<int>, double> algoritmo_greedy(vector<pair<int, int>>& puntos) {
    int n = puntos.size();
    vector<int> ruta;
    vector<bool> visitado(n, false);

    // Empezar desde la primera ciudad
    ruta.push_back(0);
    visitado[0] = true;

    while (ruta.size() < n) {
        int ciudad_actual = ruta.back();
        double min_distancia = numeric_limits<double>::infinity();
        int ciudad_proxima = -1;

        // Buscar la ciudad no visitada más cercana
        for (int i = 0; i < n; ++i) {
            if (!visitado[i] && distancia(puntos[ciudad_actual], puntos[i]) < min_distancia) {
                min_distancia = distancia(puntos[ciudad_actual], puntos[i]);
                ciudad_proxima = i;
            }
        }

        // Añadir la ciudad más cercana
        ruta.push_back(ciudad_proxima);
        visitado[ciudad_proxima] = true;
    }

    // Añadir la ciudad de inicio al final para cerrar el recorrido
    ruta.push_back(ruta[0]);

    return make_pair(ruta, largo_recorrido(ruta, puntos));
}

// Ejemplo de uso
int main() {
    srand(time(0));

    // Generación de puntos aleatorios (por ejemplo, 10 puntos)
    vector<pair<int, int>> puntos;
    for (int i = 0; i < 10; ++i) {
        puntos.push_back({rand() % 100, rand() % 100});
    }

    clock_t start_time = clock();

    // Llamar al algoritmo Greedy
    pair<vector<int>, double> resultado = algoritmo_greedy(puntos);

    vector<int> mejor_ruta = resultado.first;
    double mejor_largo = resultado.second;

    clock_t end_time = clock();

    // Imprimir el resultado
    cout << "Mejor ruta:";
    for (int i : mejor_ruta) {
        cout << " " << i;
    }
    cout << "\nDistancia total: " << mejor_largo << endl;
    cout << "Tiempo de ejecución: " << double(end_time - start_time) / CLOCKS_PER_SEC << " segundos\n";
}

File: Python/Greedy.py
import math
import random
import time

# Función para calcular la distancia entre dos puntos
def distancia(p1, p2):
    return math.sqrt((p1[0] - p2[0]) ** 2 + (p1[1] - p2[1]) ** 2)

# Función para calcular el largo de un recorrido
def largo_recorrido(ruta, puntos):
    total = 0
    for i in range(len(ruta) - 1):
        total += distancia(puntos[ruta[i]], puntos[ruta[i + 1]])
    return total

# Algoritmo Greedy
def algoritmo_greedy(puntos):
    n = len(puntos)
    ruta = [0]  # Empezar desde la primera ciudad
    visitado = [False] * n
    visitado[0] = True

    while len(ruta) < n:
        ciudad_actual = ruta[-1]
        min_distancia = float('inf')
        ciudad_proxima = -1

        # Buscar la ciudad no visitada más cercana
        for i in range(n):
            if not visitado[i] and distancia(puntos[ciudad_actual], puntos[i]) < min_distancia:
                min_distancia = distancia(puntos[ciudad_actual], puntos[i])
                ciudad_proxima = i

        ruta.append(ciudad_proxima)
        visitado[ciudad_proxima] = True

    ruta.append(ruta[0])  # Añadir la ciudad de inicio para cerrar el recorrido

    return ruta, largo_recorrido(ruta, puntos)

# Ejemplo de uso
if __name__ == '__main__':
    # Generación de puntos aleatorios
    puntos = [(random.randint(0, 100), random.randint(0, 100)) for _ in range(10)]

    start_time = time.time()
    mejor_ruta, mejor_largo = algoritmo_greedy(puntos)
    end_time = time.time()

    # Imprimir el resultado
    print("Mejor ruta:", mejor_ruta)
    print("Distancia total:", mejor_largo)
    print("Tiempo de ejecución:", end_time - start_time, "segundos")
